### PR TITLE
[Feature] Add bce loss for multilabel task

### PR DESCRIPTION
**Pull request by LXXXXR**
Hi,
In this pull request, bce loss was implemented.
Please kindly take a look. Thanks.

### CODE REVIEW

File: mmcls/models/losses/cross_entropy_loss.py [hunk 1]
```diff
@@ -18,15 +18,63 @@ def cross_entropy(pred, label, weight=None, reduction='mean', avg_factor=None):
...
+            Options are "none", "mean" and "sum". Defaults to 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. Defaults to None.
+
+    Returns:
+        torch.Tensor: The calculated loss
+    """
+    assert pred.dim() == label.dim()
+
+    loss = F.binary_cross_entropy_with_logits(pred, label, reduction='none')
```
Comment (MeowZheng) [inline mmcls/models/losses/cross_entropy_loss.py:44]: Both BCE  and  BCE_with_logits are for binary classification, but binary_cross_entropy_with_logits include sigmoid function and BCE. I think you should use `F.binary_cross_entropy`, or use an input arg to indicate use BCE or BCE_with_logits.

File: mmcls/models/losses/cross_entropy_loss.py [hunk 2]
```diff
@@ -18,15 +18,63 @@ def cross_entropy(pred, label, weight=None, reduction='mean', avg_factor=None):
...
+    loss = loss.sum(dim=1)
+
+    # apply weights and do the reduction
+    if weight is not None:
+        weight = weight.float()
+    loss = weight_reduce_loss(
+        loss, weight=weight, reduction=reduction, avg_factor=avg_factor)
+    return loss
```
Comment (MeowZheng) [inline mmcls/models/losses/cross_entropy_loss.py:52]: I thought if `reduction` is `'none'`, shape of loss must be same as input shape, but if it is not `'none'`, it should be averary or summary all elements. 
FYI. https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html?highlight=bce#torch.nn.BCELoss

File: mmcls/models/losses/cross_entropy_loss.py [hunk 3]
```diff
@@ -18,15 +18,63 @@ def cross_entropy(pred, label, weight=None, reduction='mean', avg_factor=None):
...
         super(CrossEntropyLoss, self).__init__()
+        self.use_sigmoid = use_sigmoid
         self.reduction = reduction
         self.loss_weight = loss_weight
 
-        self.cls_criterion = cross_entropy
+        if self.use_sigmoid:
+            self.cls_criterion = binary_cross_entropy
+        else:
+            self.cls_criterion = cross_entropy
```
Comment (MeowZheng) [inline mmcls/models/losses/cross_entropy_loss.py:77]: BCE is for binary classification. Cross entropy is for multi-classification.
Before BCE, prediction must be normalized by sigmoid function, so BCE_with_sigmoid just combines sigmoid and BCE. However, for multi-classification, before cross entropy, prediction must be normalized by softmax function.

**Review by MeowZheng — COMMENTED** [review]
In my opinion, you might confuse 'binary cross entropy' and 'cross entropy'.
If I misunderstood you, please point out my mistake.

File: mmcls/models/losses/cross_entropy_loss.py [hunk 4]
```diff
@@ -18,15 +18,63 @@ def cross_entropy(pred, label, weight=None, reduction='mean', avg_factor=None):
...
+                         weight=None,
+                         reduction='mean',
+                         avg_factor=None):
+    """Calculate the binary CrossEntropy loss with logits.
+
+    Args:
+        pred (torch.Tensor): The prediction with shape (N, C), where C is the
+            number of classes.
+        label (torch.Tensor): The learning label with shape (N, C), where C is
+            the number of classes.
```
Comment (MeowZheng) [inline mmcls/models/losses/cross_entropy_loss.py:32]: I thought it might be `(N, *)` instead of `(N, C)`, because it is for binary classification.  
FYI. https://pytorch.org/docs/stable/generated/torch.nn.BCELoss.html?highlight=bce#torch.nn.BCELoss

File: mmcls/models/losses/cross_entropy_loss.py [hunk 5]
```diff
@@ -18,15 +18,60 @@ def cross_entropy(pred, label, weight=None, reduction='mean', avg_factor=None):
...
+                         reduction='mean',
+                         avg_factor=None):
+    """Calculate the binary CrossEntropy loss with logits.
+
+    Args:
+        pred (torch.Tensor): The prediction with shape (N, *).
+        label (torch.Tensor): The learning label with shape (N, *).
+        weight (torch.Tensor, optional): Weight of loss. Defaults to None.
+        reduction (str, optional): The method used to reduce the loss.
+            Options are "none", "mean" and "sum". Defaults to 'mean'.
```
Comment (MeowZheng) [inline mmcls/models/losses/cross_entropy_loss.py:33]: please add "If reduction is 'none' , loss is same shape as pred and label."

**Review by MeowZheng — COMMENTED** [review]
FYI. a kind reference
https://pytorch.org/docs/stable/generated/torch.nn.MultiLabelSoftMarginLoss.html#multilabelsoftmarginloss

File: mmcls/models/losses/cross_entropy_loss.py [hunk 6]
```diff
@@ -18,15 +18,65 @@ def cross_entropy(pred, label, weight=None, reduction='mean', avg_factor=None):
     return loss
 
 
+def binary_cross_entropy(pred,
```
Comment (ycxioooong) [inline mmcls/models/losses/cross_entropy_loss.py:21]: add doc string for cross_entropy:

```python
    """Calculate the CrossEntropy loss.
    Args:
        pred (torch.Tensor): The prediction with shape (N, C), C is the number
            of classes.
        label (torch.Tensor): The learning label of the prediction.
        weight (torch.Tensor, optional): Sample-wise loss weight.
        reduction (str, optional): The method used to reduce the loss.
        avg_factor (int, optional): Average factor that is used to average
            the loss. Defaults to None.
        class_weight (list[float], optional): The weight for each class.
    Returns:
        torch.Tensor: The calculated loss
    """
```

**Review by ycxioooong — APPROVED** [review]
LGTM now

**Comment by ycxioooong** [comment]
Please fix the linting issue, then I will handle the conflicts before merging into master.
